lib/galaxy/config/__init__.py

```diff
@@ -139,40 +97,112 @@ def resolve_path(path, root):
-                break
-        else:
-            path = defaults[-1]
-
-    return resolve_path(path, root)
-
-
 def find_root(kwargs):
     root = kwargs.get('root_dir', '.')
     return root
 
 
-class Configuration(object):
+class BaseAppConfiguration(object):
+    def _set_config_base(self, config_kwargs):
+        self.sample_config_dir = os.path.join(os.path.dirname(__file__), 'sample')
+        self.config_file = find_config_file('galaxy')
+        # Parse global_conf and save the parser
+        self.global_conf = config_kwargs.get('global_conf', None)
+        self.global_conf_parser = configparser.ConfigParser()
+        if not self.config_file and self.global_conf and "__file__" in self.global_conf:
+            self.config_file = self.global_conf['__file__']
+        if self.config_file is None:
+            log.warning("No Galaxy config file found, running from current working directory: %s", os.getcwd())
+        else:
+            try:
+                self.global_conf_parser.read(self.config_file)
+            except (IOError, OSError):
+                raise
+            except Exception:
+                # Not an INI file
+                pass
+        self.config_dir = config_kwargs.get('config_dir', os.path.dirname(self.config_file or os.getcwd()))
+        self.data_dir = config_kwargs.get('data_dir', None)
+        # mutable_config_dir is intentionally not configurable. You can
+        # override individual mutable configs with config options, but they
+        # should be considered Galaxy-controlled data files and will by default
+        # just live in the data dir
+        if running_from_source:
+            if self.data_dir is None:
+                self.data_dir = os.path.join(self.root, 'database')
+            if self.config_file is None:
+                self.config_dir = os.path.join(self.root, 'config')
+            self.mutable_config_dir = self.config_dir
+            # TODO: do we still need to support ../shed_tools?
+            self.shed_tools_dir = os.path.join(self.data_dir, 'shed_tools')
+        else:
+            if self.data_dir is None:
+                self.data_dir = os.path.join(self.config_dir, 'data')
+            self.mutable_config_dir = os.path.join(self.data_dir, 'config')
+            self.shed_tools_dir = os.path.join(self.data_dir, 'shed_tools')
+        log.debug("Configuration directory is %s", self.config_dir)
+        log.debug("Data directory is %s", self.data_dir)
+        log.debug("Mutable config directory is %s", self.mutable_config_dir)
+
+    def _in_mutable_config_dir(self, path):
+        return os.path.join(self.mutable_config_dir, path)
+
+    def _in_config_dir(self, path):
+        return os.path.join(self.config_dir, path)
+
+    def _in_sample_dir(self, path):
+        return os.path.join(self.sample_config_dir, path)
+
+    def _parse_config_file_options(self, defaults, listify_defaults, config_kwargs):
+        for var, defaults in defaults.items():
```

Comment:
Probably it's not a bug, but reusing `defaults` variable name as loop variable is surely confusing.

Reply:
Yes!! I noticed it and hated it on first site.. I'll submit a quick PR (didn't want to trigger a code review over such a minor thing - but since you mention this.. so thanks :)

Reply:
@ic4f thanks!

Reply:
@nsoranzo How about defaults >> vals? 
```
for var, vals in defaults.items():
    if config_kwargs.get(var, None) is not None:
        path = config_kwargs.get(var)
        setattr(self, var + '_set', True)
    else:
        for val in vals:
            if os.path.exists(resolve_path(val, self.root)):
                path = val 
                break
        else:
            path = vals[-1]
        setattr(self, var + '_set', False)
    setattr(self, var, resolve_path(path, self.root))
```

Reply:
Maybe `values`/`value` to make it a bit more different from `var`.